server/testnet.go

```diff
@@ -0,0 +1,240 @@
+/*
+ * Ported from Tendermint
+ */
+package server
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	cfg "github.com/tendermint/tendermint/config"
+	gaiacfg "github.com/cosmos/cosmos-sdk/config"
+	"github.com/tendermint/tendermint/p2p"
+	"github.com/tendermint/tendermint/types"
+	pvm "github.com/tendermint/tendermint/types/priv_validator"
+	cmn "github.com/tendermint/tmlibs/common"
+	"github.com/cosmos/cosmos-sdk/wire"
+)
+
+var (
+	nValidators    int
+	nNonValidators int
+	outputDir      string
+	nodeDirPrefix  string
+
+	populatePersistentPeers bool
+	hostnamePrefix          string
+	startingIPAddress       string
+	p2pPort                 int
+)
+
+const (
+	nodeDirPerm = 0755
+)
+
+func init() {
+}
+
+// get cmd to initialize all files for tendermint testnet and application
+func TestnetFilesCmd(ctx *Context, cdc *wire.Codec, appInit AppInit) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "testnet",
+		Short: "Initialize files for a Gaiad testnet",
+		Long: `testnet will create "v" + "n" number of directories and populate each with
+necessary files (private validator, genesis, config, etc.).
+
+Note, strict routability for addresses is turned off in the config file.
+
+Optionally, it will fill in persistent_peers list in config file using either hostnames or IPs.
+
+Example:
+
+	gaiad testnet --v 4 --o ./output --populate-persistent-peers --starting-ip-address 192.168.10.2
+	`,
+		RunE: func(_ *cobra.Command, _ []string) error {
+			config := ctx.Config
+			genVals := make([]types.GenesisValidator, nValidators)
+
+			// Generate private key, node ID, initial transaction
+			for i := 0; i < nValidators; i++ {
+				nodeDirName := cmn.Fmt("%s%d", nodeDirPrefix, i)
+				nodeDir := filepath.Join(outputDir, nodeDirName, "gaiad")
+				clientDir := filepath.Join(outputDir, nodeDirName, "gaiacli")
+				gentxsDir := filepath.Join(outputDir, "gentxs")
+				config.SetRoot(nodeDir)
+
+				err := os.MkdirAll(filepath.Join(nodeDir, "config"), nodeDirPerm)
+				if err != nil {
+					_ = os.RemoveAll(outputDir)
```

Comment:
why are we skipping this input? (for below RemoveAll as well)

Reply:
It's a "best-effort" cleanup (copied from the Tendermint code). We already have an error message, so while we're quitting with that message, we're trying to remove the half-baked configuration folders that we created. If it succeeds, good, if not, we still want to present the higher priority error message to the user. (And let the user do the cleanup.)